engine/resources/hetzner.go

```diff
@@ -0,0 +1,710 @@
+	}
+	if obj.stateConverged() {
+		return checkOk, nil
+	}
+	// Request changes if the server is not in the intended state yet.
+	// TODO: add permissions (poweron, poweroff, delete)
+	checkOk = false
+	if !apply {
+		return false, nil
+	}
+	switch obj.State {
+	case StateRunning:
+		obj.init.Logf("requesting server poweron")
+		if _, _, err := obj.client.Server.Poweron(ctx, obj.server); err != nil {
+			return false, errwrap.Wrapf(err, "client.server.poweron failed")
+		}
+	case StateOff:
+		obj.init.Logf("requesting server poweroff")
+		if _, _, err := obj.client.Server.Poweroff(ctx, obj.server); err != nil {
+			return false, errwrap.Wrapf(err, "client.server.poweroff failed")
+		}
+	case StateAbsent:
+		obj.init.Logf("requesting server deletion")
+		if _, err := obj.client.Server.Delete(ctx, obj.server); err != nil {
+			return false, errwrap.Wrapf(err, "client.server.delete failed")
+		}
+	}
+	// Wait until the requested changes are confirmed. Error otherwise.
+	if err := obj.getServerUpdate(ctx); err != nil {
+		return false, errwrap.Wrapf(err, "getServerUpdate failed")
+	}
+	if err := obj.waitForConvergedState(ctx); err != nil {
+		return false, errwrap.Wrapf(err, "waitForConvergedState failed")
+	}
+	return checkOk, nil
+}
+
+// checkApplyServerRebuild rebuilds the server instance as needed,
+// to ensure that the resource matches the provided specifications.
+// TODO: consider using client.server.rebuild(image) for faster image rebuild.
+func (obj *HetznerRes) checkApplyServerRebuild(ctx context.Context, apply bool) (bool, error) {
+	if obj.init.Debug {
+		obj.init.Logf("checkApplyServerRebuild(apply: %t)", apply)
+	}
+	specsMatch, err := obj.cmpServerSpecs()
+	if err != nil {
+		return false, errwrap.Wrapf(err, "cmpServerSpecs failed")
+	}
+	var checkOk = true
+	if specsMatch {
+		return true, nil
+	}
+	checkOk = false
+	if !apply {
+		return false, nil
+	}
+	// TODO: add rebuild permission
+	obj.init.Logf("requesting server rebuild")
+	if err := obj.rebuildServer(ctx); err != nil {
+		return false, errwrap.Wrapf(err, "rebuildServer failed")
+	}
+	return checkOk, nil
+}
+
+// checkApplyRescueMode enables/disables rescue mode for the given keyset,
+// to ensure that rescue mode is available over SSH when needed.
+// TODO: add support for password authentication, now only ssh.
+// TODO: add checkapply for rescue mode specs when already enabled?
+// Right now you must fully disable and re-enable to update specs.
+func (obj *HetznerRes) checkApplyRescueMode(ctx context.Context, apply bool) (bool, error) {
+	if obj.init.Debug {
+		obj.init.Logf("checkApplyRescueMode(apply: %t)", apply)
+	}
+	// Check if rescue mode is already in the intended configuration.
+	var checkOk = true
+	if obj.server.RescueEnabled == obj.rescueEnable {
+		return true, nil
+	}
+	checkOk = false
+	if !apply {
+		return false, nil
+	}
+	// Otherwise, apply the necessary changes.
+	switch obj.rescueEnable {
```

Comment:
why a switch?

Reply:
The if-else seemed less readable, but it's no hard preference. Changed it again to separate if statements.